Source/init.cpp

```diff
@@ -98,25 +101,24 @@ void init_archives()
 #endif
 	WCloseFile(fh);
 #ifdef SPAWN
-	patch_rt_mpq = init_test_access(patch_rt_mpq_path, "patch_sh.mpq", "DiabloSpawn", 2000, FS_PC);
+	patch_rt_mpq = init_test_access(&patch_rt_mpq_path, "patch_sh.mpq", "DiabloSpawn", 2000, FS_PC);
 #else
-	patch_rt_mpq = init_test_access(patch_rt_mpq_path, "patch_rt.mpq", "DiabloInstall", 2000, FS_PC);
+	patch_rt_mpq = init_test_access(&patch_rt_mpq_path, "patch_rt.mpq", "DiabloInstall", 2000, FS_PC);
 #endif
 }
 
-HANDLE init_test_access(char *mpq_path, char *mpq_name, char *reg_loc, int flags, int fs)
+HANDLE init_test_access(char **mpq_path, char *mpq_name, char *reg_loc, int flags, int fs)
 {
-	char Buffer[2][MAX_PATH];
-	char *sdlPath;
-	HANDLE archive;
-
-	GetBasePath(Buffer[0], MAX_PATH);
-	GetPrefPath(Buffer[1], MAX_PATH);
-
-	for (int i = 0; i < 2; i++) {
-		snprintf(mpq_path, MAX_PATH, "%s%s", Buffer[i], mpq_name);
-		if (SFileOpenArchive(mpq_path, flags, MPQ_FLAG_READ_ONLY, &archive)) {
-			SFileSetBasePath(Buffer[i]);
+	std::array<std::string, 2> mpqPaths = {
+		GetBasePath() + mpq_name,
+		GetPrefPath() + mpq_name,
+	};
+
+	for (const std::string &path : mpqPaths) {
+		HANDLE archive;
+		if (SFileOpenArchive(path.c_str(), flags, MPQ_FLAG_READ_ONLY, &archive)) {
+			*mpq_path = new char[path.size() + 1];
+			strcpy(*mpq_path, path.c_str());

```

Comment:
This is a problem if the string is MAX_PATH or more.

Reply:
Shouldn't be? `mpq_path` is allocated exactly to the size of `path` (+1 for `\0)

Reply:
I see, the MPQ library itself internally allocated a 256 char buffer.
It doesn't as of the latest.

Reply:
Is mpq_path zero filled on init on all platforms?

Reply:
It's not zero-filled but `strcpy` copies the `\0` is well

Reply:
Good point :)